File: docs/examples/getting-started/python/server-02.py
from dataclasses import dataclass


@dataclass
class Expense:
    amount: int
    description: str
    submitted_by: str


EXPENSES = {
    1: Expense(500, "coffee", "alice@example.com"),
    2: Expense(5000, "software", "alice@example.com"),
    3: Expense(50000, "flight", "bhavik@example.com"),
}
# expenses code

from oso import Oso

OSO = Oso()

# server code

from http.server import HTTPServer, BaseHTTPRequestHandler


class MyRequestHandler(BaseHTTPRequestHandler):
    def _respond(self, msg):
        self.send_response(200)
        self.end_headers()
        self.wfile.write(msg.encode())

    def do_GET(self):
        try:
            _, resource_type, resource_id = self.path.split("/")
            resource = EXPENSES[int(resource_id)]
            self._respond(str(resource))
        except (KeyError, ValueError) as e:
            self._respond("Not Found!")


server_address = ("", 5050)
httpd = HTTPServer(server_address, MyRequestHandler)
httpd.serve_forever()

File: docs/examples/quickstart/ruby/server-02.rb
require "oso"
require "webrick"

class Expense
  attr_reader :amount, :description, :submitted_by

  def initialize(amount, description, submitted_by)
    @amount = amount
    @description = description
    @submitted_by = submitted_by
  end
end

EXPENSES = {
  1 => Expense.new(500,   "coffee",   "alice@example.com"),
  2 => Expense.new(5000,  "software", "alice@example.com"),
  3 => Expense.new(50000, "flight",   "bhavik@example.com"),
}

OSO ||= Oso.new

server = WEBrick::HTTPServer.new Port: 5050
server.mount_proc "/" do |req, res|
  _, resource_type, resource_id = req.path.split("/")
  resource = EXPENSES[resource_id.to_i]

  if resource_type != "expenses" || resource.nil?
    res.body = "Not Found!"
  else
    res.body = resource.inspect
  end
end
server.start
